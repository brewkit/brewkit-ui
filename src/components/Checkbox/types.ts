import * as React from 'react';
import { Theme } from '@material-ui/core/styles/createMuiTheme';
import { CheckboxProps } from '@material-ui/core/Checkbox';


/**
 * Declare any theme variables we want to make available.
 */
declare module '@material-ui/core/styles/createMuiTheme' {
    interface Theme {
        BkCheckbox: {
            controlBase: React.CSSProperties,
            uncheckedControl: React.CSSProperties,
            checkedControl: React.CSSProperties,
        },
    }
    // allow configuration using `createMuiTheme`
    interface ThemeOptions {
        BkCheckbox?: {
            controlBase?: React.CSSProperties,
            uncheckedControl?: React.CSSProperties,
            checkedControl?: React.CSSProperties,
        },
    }
}


export type Props = CheckboxProps & {

    /**
     * The color of the component. It supports those theme colors that make sense for this component.
     * @default "primary"
     */
    color?: 'default' | 'primary' | 'secondary',

    /**
     * @ignore
     */
    classes?: Record<string, any>,

};
