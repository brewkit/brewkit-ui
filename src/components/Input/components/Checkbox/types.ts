import { Props as InputBaseProps } from '../InputBase/types';


export type Props = Omit<InputBaseProps,
    'as'
    | 'after'
    | 'before'
    | 'children'
    | 'cols'
    | 'defaultValue'
    | 'multiple'
    | 'placeholder'
    | 'rows'
    | 'type'
    | 'value'
>;
