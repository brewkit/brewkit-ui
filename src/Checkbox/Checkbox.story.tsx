import * as React from 'react';
import FormControlLabel from '@material-ui/core/FormControlLabel';

/**
 * This is an annoying hack to get Storybook to properly parse the documentation when the underlying component
 * is wrapped in a HOC. Expect to do/see this on every component with MUI theming options.
 */
import { Checkbox as Component } from './Checkbox';
const Checkbox = Component;

export default {
    component: Component,
    title: 'Inputs/Checkbox',
};

const defaultArgs = {
    disabled: false,
    checked: false,
    indeterminate: false,
};

export const Sandbox = (args: any): React.ReactElement => (
    <Checkbox {...args} />
);
Sandbox.args = {
    ...defaultArgs,
    color: 'primary',
};

export const General = (): React.ReactElement => (
    <div>
        <Checkbox
            defaultChecked
            inputProps={{ 'aria-label': 'secondary checkbox' }}
        />
        <Checkbox
            color="secondary"
            inputProps={{ 'aria-label': 'uncontrolled-checkbox' }}
        />
        <Checkbox disabled inputProps={{ 'aria-label': 'disabled checkbox' }} />
        <Checkbox
            checked
            disabled
            inputProps={{ 'aria-label': 'disabled checked checkbox' }}
        />
        <Checkbox
            defaultChecked
            indeterminate
            inputProps={{ 'aria-label': 'indeterminate checkbox' }}
        />
        <Checkbox
            color="default"
            defaultChecked
            inputProps={{ 'aria-label': 'checkbox with default color' }}
        />
        <Checkbox
            defaultChecked
            inputProps={{ 'aria-label': 'checkbox with small size' }}
            size="small"
        />
    </div>
);

export const WithLabel = (): React.ReactElement => (
    <div>
        <FormControlLabel
            control={<Checkbox defaultChecked name="checkedA" />}
            label="Primary"
        />
        <FormControlLabel
            control={
                <Checkbox color="secondary" defaultChecked name="checkedA" />
            }
            label="Secondary"
        />
        <FormControlLabel
            control={<Checkbox defaultChecked indeterminate name="checkedA" />}
            label="Indeterminate"
        />
        <FormControlLabel
            control={<Checkbox defaultChecked name="checkedA" />}
            disabled
            label="Disabled Checked"
        />
        <FormControlLabel
            control={<Checkbox name="checkedA" />}
            disabled
            label="Disabled Unchecked"
        />
    </div>
);

export const LabelPlacement = (): React.ReactElement => (
    <div>
        <FormControlLabel
            control={<Checkbox />}
            label="Top"
            labelPlacement="top"
            value="top"
        />
        <FormControlLabel
            control={<Checkbox />}
            label="Start"
            labelPlacement="start"
            value="start"
        />
        <FormControlLabel
            control={<Checkbox />}
            label="Bottom"
            labelPlacement="bottom"
            value="bottom"
        />
        <FormControlLabel
            control={<Checkbox />}
            label="End"
            labelPlacement="end"
            value="end"
        />
    </div>
);
