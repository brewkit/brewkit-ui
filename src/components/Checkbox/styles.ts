import { Theme } from '@material-ui/core/styles/createMuiTheme';
import blueGrey from '@material-ui/core/colors/blueGrey';
import { merge as _merge } from 'lodash';


const styles = (theme: Theme): any => ({


    controlBase: _merge({
        height: '2rem',
        width: '2rem',
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        borderWidth: '2px',
        borderStyle: 'solid',
        borderRadius: '4px',
        transition: 'all 0.3s',
    }, theme?.BkCheckbox?.controlBase),


    uncheckedControl: _merge({
        borderColor: blueGrey[100],
    }, theme?.BkCheckbox?.uncheckedControl),


    checkedControl: _merge({
        color: '#fff',
        backgroundColor: theme?.palette?.primary?.main,
        borderColor: theme?.palette?.primary?.main,
    }, theme?.BkCheckbox?.checkedControl),


});


export default styles;
