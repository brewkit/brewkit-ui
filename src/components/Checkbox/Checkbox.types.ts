
import * as React from 'react';
import { CheckboxProps } from '@material-ui/core/Checkbox';


/**
 * Declare any theme variables we want to make available.
 */
declare module '@material-ui/core/styles/createTheme' {
    interface Theme {
        BkCheckbox: {
            controlBase: React.CSSProperties,
            uncheckedControl: React.CSSProperties,
            checkedControl: React.CSSProperties,
        },
    }
    // allow configuration using `createTheme`
    interface ThemeOptions {
        BkCheckbox?: {
            controlBase?: React.CSSProperties,
            uncheckedControl?: React.CSSProperties,
            checkedControl?: React.CSSProperties,
        },
    }
}


export type Props = CheckboxProps & {

    /**
     * The color of the component. It supports those theme colors that make sense for this component.
     * @default "primary"
     */
    color?: 'default' | 'primary' | 'secondary',

    /**
     * @ignore
     */
    classes?: Record<string, any>,

};
