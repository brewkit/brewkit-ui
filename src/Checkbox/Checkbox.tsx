import React from 'react';
import clsx from 'clsx';
import { motion, AnimatePresence } from 'framer-motion';
import { merge } from 'lodash';
import MUICheckbox, {
    CheckboxClassKey,
    CheckboxProps as MUICheckboxProps,
} from '@material-ui/core/Checkbox';
import Check from '@material-ui/icons/Check';
import Remove from '@material-ui/icons/Remove';
import { ClassNameMap } from '@material-ui/core/styles/withStyles';
import { useStyles } from './Checkbox.styles';
import { withoutKeys } from '../utils/withoutKeys';

export type BkCheckboxProps = MUICheckboxProps & {
    /**
     * @ignore
     */
    classes?: ClassNameMap<
        CheckboxClassKey | 'controlBase' | 'uncheckedControl' | 'checkedControl'
    >;
};

const Checkbox = React.forwardRef(
    (props: BkCheckboxProps, ref: React.Ref<any>): React.ReactElement => {
        const {
            color = 'primary',
            classes: userClasses,
            className,
            ...otherProps
        } = props;
        const classes = useStyles();
        const mergedClasses = merge(
            withoutKeys(classes, [
                'controlBase',
                'checkedControl',
                'uncheckedControl',
            ]),
            userClasses,
        );

        /**
         * configure our Framer animation
         */
        const fadeAnim = {
            initial: { opacity: 0 },
            animate: { opacity: 1 },
            exit: { opacity: 0 },
        };

        /**
         * the control that will be used if checked
         */
        const CheckedIcon = (
            <div
                className={clsx(
                    classes?.controlBase,
                    classes?.checkedControl,
                    userClasses?.controlBase,
                    userClasses?.checkedControl,
                )}
                data-testid="checkedIcon">
                <Check component={motion.svg} key="checkedIcon" {...fadeAnim} />
            </div>
        );

        /**
         * the control that will be used if unchecked
         */
        const UncheckedIcon = (
            <div
                className={clsx(
                    classes?.controlBase,
                    classes?.uncheckedControl,
                    userClasses?.controlBase,
                    userClasses?.uncheckedControl,
                )}
                data-testid="uncheckedIcon">
                <motion.div key="icon" {...fadeAnim} />
            </div>
        );

        /**
         * the control that will be used if indeterminate
         */
        const IndeterminateIcon = (
            <div
                className={clsx(classes?.controlBase, classes?.checkedControl)}>
                <Remove
                    component={motion.svg}
                    key="indeterminateIcon"
                    {...fadeAnim}
                />
            </div>
        );

        return (
            <AnimatePresence>
                <MUICheckbox
                    color={color}
                    classes={mergedClasses}
                    checkedIcon={CheckedIcon}
                    icon={UncheckedIcon}
                    indeterminateIcon={IndeterminateIcon}
                    ref={ref}
                    {...otherProps}
                />
            </AnimatePresence>
        );
    },
);

Checkbox.displayName = 'Checkbox';

export default Checkbox;
