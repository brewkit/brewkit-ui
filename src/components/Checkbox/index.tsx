import * as React from 'react';
import clsx from 'clsx';
import { motion, AnimatePresence } from 'framer-motion';
import { withStyles } from '@material-ui/core/styles';
import MuiCheckbox from '@material-ui/core/Checkbox';
import Check from '@material-ui/icons/Check';
import Remove from '@material-ui/icons/Remove';
import { Props } from './types';
import styles from './styles';


/**
 * Checkboxes allow the user to select one or more items from a set.
 *
 * [Material-UI Docs](https://material-ui.com/components/checkboxes/)
 *
 *
 * ## Differences in Brewkit:
 *
 * - Changes the default controls for a different look and feel.
 *
 *   > The change in controls can lead to some mismatching in colors when `color="secondary"`, so be prepared to
 *   have to account for that in your theme's style overrides.
 *
 * ## Additional Classes
 * - `BkCheckbox.controlBase`: Styles applied to all controls
 * - `BkCheckbox.checkedControl`: Additional styles applied to the checked control
 * - `BkCheckbox.uncheckedControl`: Additional styles applied to the unchecked control
 */
export const Checkbox = React.forwardRef(({
    classes,
    className,
    ...otherProps
}: Props, ref: React.Ref<any>): React.ReactElement => {


    /**
     * merge our custom classes into the ones passed to the component
     */
    const checkboxClasses = clsx(
        classes?.root,
        className,
    );


    /**
     * configure our Framer animation
     */
    const fadeAnim = {
        initial: { opacity: 0 },
        animate: { opacity: 1 },
        exit: { opacity: 0 },
    };


    /**
     * the control that will be used if checked
     */
    const CheckedIcon = (
        <div className={clsx(classes?.controlBase, classes?.checkedControl)}>
            <Check component={motion.svg} key="checkedIcon" {...fadeAnim} />
        </div>
    );


    /**
     * the control that will be used if unchecked
     */
    const UncheckedIcon = (
        <div className={clsx(classes?.controlBase, classes?.uncheckedControl)}>
            <motion.div key="icon" {...fadeAnim} />
        </div>
    );


    /**
     * the control that will be used if indeterminate
     */
    const IndeterminateIcon = (
        <div className={clsx(classes?.controlBase, classes?.checkedControl)}>
            <Remove component={motion.svg} key="indeterminateIcon" {...fadeAnim} />
        </div>
    );


    return (
        <AnimatePresence>
            <MuiCheckbox
                checkedIcon={CheckedIcon}
                className={checkboxClasses}
                icon={UncheckedIcon}
                indeterminateIcon={IndeterminateIcon}
                ref={ref}
                {...otherProps}
            />
        </AnimatePresence>
    );


});


Checkbox.displayName = 'Checkbox';


export default withStyles(styles)(Checkbox);
