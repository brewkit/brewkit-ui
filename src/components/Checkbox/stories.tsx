import * as React from 'react';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import { select, boolean } from '@storybook/addon-knobs';
import { withStyles } from '@material-ui/core/styles';


/**
 * This is an annoying hack to get Storybook to properly parse the documentation when the underlying component
 * is wrapped in a HOC. Expect to do/see this on every component with MUI theming options.
 */
import { Checkbox as Component } from './index';
import styles from './styles';
const Checkbox = withStyles(styles)(Component)


export default {
    component: Component,
    title: 'Inputs/Checkbox',
};


export const Sandbox = (): React.ReactElement => {

    const color = select('color', ['default', 'primary', 'secondary'], 'default');
    const disabled = boolean('disabled', false);

    return (
        <Checkbox
            color={color}
            disabled={disabled}
        />
    );

};


export const General = (): React.ReactElement => {

    return (
        <div>
            <Checkbox
                defaultChecked
                inputProps={{ 'aria-label': 'secondary checkbox' }}
            />
            <Checkbox color="secondary" inputProps={{ 'aria-label': 'uncontrolled-checkbox' }} />
            <Checkbox disabled inputProps={{ 'aria-label': 'disabled checkbox' }} />
            <Checkbox disabled checked inputProps={{ 'aria-label': 'disabled checked checkbox' }} />
            <Checkbox
                defaultChecked
                indeterminate
                inputProps={{ 'aria-label': 'indeterminate checkbox' }}
            />
            <Checkbox
                defaultChecked
                color="default"
                inputProps={{ 'aria-label': 'checkbox with default color' }}
            />
            <Checkbox
                defaultChecked
                size="small"
                inputProps={{ 'aria-label': 'checkbox with small size' }}
            />
        </div>
    );

}


export const WithLabel = (): React.ReactElement => {

    return (
        <div>
            <FormControlLabel
                control={<Checkbox defaultChecked name="checkedA" />}
                label="Primary"
            />
            <FormControlLabel
                control={<Checkbox defaultChecked name="checkedA" color="secondary" />}
                label="Secondary"
            />
            <FormControlLabel
                control={<Checkbox defaultChecked name="checkedA" indeterminate />}
                label="Indeterminate"
            />
            <FormControlLabel
                control={<Checkbox defaultChecked name="checkedA" />}
                disabled
                label="Disabled Checked"
            />
            <FormControlLabel
                control={<Checkbox name="checkedA" />}
                disabled
                label="Disabled Unchecked"
            />
        </div>
    );

}


export const LabelPlacement = (): React.ReactElement => {

    return (
        <div>
            <FormControlLabel
                value="top"
                control={<Checkbox />}
                label="Top"
                labelPlacement="top"
            />
            <FormControlLabel
                value="start"
                control={<Checkbox />}
                label="Start"
                labelPlacement="start"
            />
            <FormControlLabel
                value="bottom"
                control={<Checkbox />}
                label="Bottom"
                labelPlacement="bottom"
            />
            <FormControlLabel
                value="end"
                control={<Checkbox />}
                label="End"
                labelPlacement="end"
            />
        </div>
    );

}
