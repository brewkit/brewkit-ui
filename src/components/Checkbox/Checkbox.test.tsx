// @ts-nocheck
import * as React from 'react';
import { create } from 'react-test-renderer';
import Checkbox from '.';

it('Renders checkbox component', () => {
    const comp = create(
        <Checkbox />
    ).toJSON();

    expect(comp).toMatchSnapshot();
});
