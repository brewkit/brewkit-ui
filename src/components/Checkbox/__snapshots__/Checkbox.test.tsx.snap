// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders checkbox component 1`] = `
<span
  aria-disabled={false}
  className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-4 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
  onBlur={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={null}
>
  <span
    className="MuiIconButton-label"
  >
    <input
      className="PrivateSwitchBase-input-7"
      data-indeterminate={false}
      onChange={[Function]}
      type="checkbox"
    />
    <div
      className="Checkbox-controlBase-1 Checkbox-uncheckedControl-2"
    >
      <div
        style={
          Object {
            "opacity": 0,
          }
        }
      />
    </div>
  </span>
</span>
`;
